### ISSUE TITLE
Authorization and multi-tenant user management system

### ISSUE THREAD

**Issue by nathanielrindlaub** [tnc-ca-geo/animl-api#37]
The goal is to implement an auth and user management system that supports the following: 

- Separate, siloed “projects”, in which read access to images and all other permissions are limited to users who are members of of that project
- Tiered user roles, to further limit/grant permissions within within each project:
	- _Product Owner/Super user/admin_: Natty/Falk created 2 new projects: one for SCI biosecurity, and one for a large carnivore study on the Dangermond preserve. They can create and add users to those projects and set those users permission levels. 
	- _Project Manager/Project admin_: Juli, the NPS Biosecurity manager, (and probably Natty & Falk as well) are admins for the SCI biosecurity project, they can add users, configure inference pipelines and other automation rules, see all images that belong to that project, and edit labels. They can create and edit views. If the data model requires that cameras or base stations be registered to projects, the project admins can do this. They can also upload images directly from their computers. 
	- _Project Contributor_: Miles added this one, but I am not sure we need it. Essentially A role permitting image/asset write access.
	- _Project Member/Project reviewer_: Will is an NPS biosecurity intern for the summer, and part of his job is to help review and edit labels - he will need read access to all images within a project and write access to the labels, and he can create views to help him with his review workflow. He can not edit inference pipelines or create new users. 
	- _Project observer_: I’m not 100% sure we need this one, but basically a read-only role. Perhaps this is useful for ecologists or researchers outside of our organization who want to consume the reviewed, validated data to build a population model but we don’t necessarily want them editing any labels. 

We will utilize Cognito groups to organize both the projects and roles. 
- Each group name will contain both the project name/id and role, e.g.: `animl/sci_biosecurity/project_owner`
- When a user is authenticated by Cognito, the returned ID token will include a list of all groups (project + role combinations) they are a member of. 
- The front end parses out the projects and roles the user is a member of, builds list of projects in state to allow a user to navigate between projects so that they know what project they're "logged into" and acting on behalf of. 
- Front end will also show/hide/disable certain functionality based on role. 
- All requests to the API will contain the ID token, but we also need to figure out a way to pass in the current group the user is acting on behalf of... perhaps a custom header? 
- The animl-api resolvers (at the model level) will be responsible for roll gatekeeping operations (making sure the user's role has permission to perform the action)
- The siloing of data by project will happen at the DB query/filtering level: every entity/record in the DB will be associated with a project, and we will simply append "project = usersCurrentProjectId" to all queries. 
- This requires associating all new entities with a project when they're created.
	- For manually created records, like a view or automation rule, the project would be whatever project the user is currently "logged into"
	- For automatically created image records, users must first "register" that particular camera's serial number with their project. They would use the UI do request to add a new camera, the API would then check to make sure that serial number isn't already associated with a different project, if not it associates it with the user's current project, and all new images coming into the system from that camera will be tagged with that project.
	- Future ideas: implement the ability for projects to "release"/"unregister" a serial number in case the camera needs to be moved to a different project, freeing it up for a different project to grab it
	- also maybe implement a way for different projects to request access to those cameras images or some subset of them.... so there's a primary owner of each image but maybe a list of other projects that have read-only access or something? not sure how that would work exactly. 
	- Not sure we need to associate projects w/ label records. Kind of depends if an image can be viewed or owned by more than one project. If no, then we don’t really need to, if yes, then maybe we do, as different projects might have different labeling goals/needs?

**Comment by nathanielrindlaub** [comment]
Updated approach: I made some adjustments to the DB schema after adding a `Projects` collection and sitting down to try to implement the approach outlined above, and I wanted to document those decisions here.

TL;DR, for both `MLModels` and `Cameras`, we give them their own "source" collections, which represent the entities independently of any specific project (i.e., cameras and ML models can exist without being associated with a project). However, in order to support project-level associations/relationships and project-level configurations for them, we embed somewhat-duplicative representations of the entities within `Projects` documents that contain more project-specific configuration data. This pattern might sound odd but it's actually recommended in certain many-to-many instances in [MongoDB's own documentation](https://youtu.be/QzBI0_fxMoc) (the example they give is a solution for representing an item in a store both within its own collection of `Items` and also embedded within a users shopping cart). 

The reason I went this route was because the relationship between `Projects`, `Cameras`, and `Deployments` is complicated, especially if we want to support releasing and re-assigning cameras to various projects over their lifetime. In short, projects have a many-to-many relationship with cameras, cameras have a one-to-many relationship with deployments, but deployments are specific to projects, and are really just a tool for organizing the images that that project owns that were generated by a given camera. The solution I came up with was to use the `Cameras` collection to hold the "source" camera records, which contains make, model, and a list of current and previous associations with projects, and to embed an array of `CameraConfigs` in each project, which contains their project-specific deployments:

Cameras.js (source record)
```javascript
let ProjectRegistrationSchema = new Schema({
  project: { type: String, default: 'default_project', required: true },
  active: { type: Boolean, required: true },
});

let CameraSchema = new Schema({
  _id: { type: String, required: true },
  make: { type: String, default: 'unknown', required: true },
  model: { type: String },
  projRegistrations: { type: [ProjectRegistrationSchema] },
});
```

Projects.js (w/ embedded camera config)
```javascript
...
let CameraConfigSchema = new Schema({
  _id: { type: String, required: true },  // _id is camera serial number
  deployments: { type: [DeploymentSchema]},
});

let ProjectSchema = new Schema({
  ...
  cameras: { type: [CameraConfigSchema]}, 
  ...
});
...
```

Because we also want to support project-level (technically automation-rule-level) configuration settings for ML models, we also took a similar approach with model records. Now each model has a "source" record in a stand-alone `MLModels` collection that includes the model version numbers, category/class mappings, and default confidence thresholds. When a user creates a new automation rule requesting inference, however, the form is drawing on the "source" model records for default values but allows users to override them (adjust confidence thresholds for each class, disable certain classes entirely). Those config settings get baked into the the newly created automation rule and are used when inference is requested.
